README.md
台灣新聞拆拆樂的用來蒐集或處理大眾感興趣的資訊，目前著重在新聞和市場情報，希望能藉由這個工具讓其他人變出各種好玩的衍生應用，目前提供這些功能：

* 將新聞網頁拆出標題，內文，報導時間，記者
* 關鍵字查新聞，並且可接續拆新聞或新聞數
* 蒐集三大法人，信用交易，借券賣出，股權分散表等市場資訊

## 為什麼做這個？

最早的開發動機，是希望能做個工具能快速找到最新的凶宅，用來輔助我的另一個專案[鄉民風水師](https://geomancer.tacosync.com/)，實現資料自動更新，所以做了拆站工具和搜尋爬蟲。做了一段時間也開始有話題性，就修了一些大家反應的問題。然後希望維護這專案的負擔不要太重，又寫了一些能發財的東西，然後就變成現在的樣子。如果想了解更多以前的事，可以參考[版本異動紀錄](docs/changelog.md)。

如果你是來找發財工具，另一個乏人問津的專案[群益證券聽牌機](https://github.com/virus-warnning/skcom)也歡迎來關注一下。 

## 怎麼用？

**系統需求**

* OS: 不限
* Python: 3.5 ~ 3.7

**安裝**

```sh
# Windows (只有 Python 3.x 的環境)
pip install twnews

# macOS, Linux (Python 2.x 與 3.x 並存的環境)
pip3 install twnews
```

**體驗**

為了方便有緣人使用，開發了一些 console 工具讓人可以不用寫程式就立即體驗

例如朝聖一則凶宅新聞：

```bash
python3 -m twnews soup https://www.setn.com/News.aspx?NewsID=490688&From=Search&Key=%E5%8F%B0%E4%B8%AD%E6%9C%80%E7%8C%9B%E5%85%87%E5%AE%85%EF%BC%81%E9%80%A36%E5%90%8D%E6%88%BF%E5%AE%A2%E5%90%8C%E4%BD%8D%E7%BD%AE%E4%B8%8A%E5%90%8A
---------------------------------------------------------------------------
原始路徑: https://www.setn.com/News.aspx?NewsID=490688
最終路徑: https://www.setn.com/m/news.aspx?newsid=490688
頻道: setn
標題: 堪稱台中最猛兇宅！連續6名房客上吊　竟都死在相同位置
日期: 2019-02-07 20:00:00
記者: 林盈君
內文:
記者林盈君／台中報導許多人不論是租屋、購屋，都擔心會碰上兇宅，有位職業為房仲的網友，就在PTT上，分享他曾經遇過的詭異事件！
他說約在幾年前，接到一個同行介紹的案子，是一間位於台中沙鹿的頂樓老公寓，當時他看到房屋價格時，簡直低到讓他不敢相信，詢問過
屋主之後，屋主表示屋內曾有人上吊死亡，而且是「連續6名房客」，且竟然都在「同一位置」選擇輕生！▲該公寓曾發生6名租客上吊事件
。（圖／示意圖／翻攝pixabay）據悉，幾年前曾有一名房客，在屋內上吊自殺，屋主之後請了法師來超渡誦經，就低價再次租給其他房客
，不料6個月後，該房客卻遲遲沒有交租金，就連簡訊電話都無回應，他只好請警察協助開門，沒想到進到屋內就發現，房客已經上吊死亡
，而且死亡的位置，跟第一個房客上吊的地方「一模一樣」，實在令人毛骨悚然。▲許多人認為這根本是「抓交替」。（圖／示意圖／翻攝
pixabay）在那之後，屋主連續租給其他幾位房客，但卻頻頻發生意外，6位租客全都選擇在屋內上吊自殺，上吊的位置也全部一模一樣，
這樣詭異的情況，也讓轄區員警懷疑，屋主是不是殺害租客後，用上吊來掩蓋罪行，讓屋主相當崩潰！屋主曾無奈表示「1、2個就算了，是
連續6個租客你知道嗎？」、「這6個租客我根本不知道為什麼他們最後都上吊死了，而且上吊的位置全都一模一樣」。更加詭異的是，連續
6名房客上吊身亡後，屋主有2年多都沒再把房子出租，屋內空無一人，但對面鄰居卻跟他表示，看到有人進出他的房子，樓下的住戶也說，
上面常常傳來很用力踩地板、拖著椅子及摔東西的聲音，甚至連一樓的住戶，都說看到房子的陽台上有人，但等到屋主拿鑰匙進屋查看時，
發現從門口到陽台都是厚厚的灰塵，完全沒有任何人走動的跡象。因為發生過命案，再加上接連詭異的情況，讓對面鄰居和樓下住戶都陸續
搬走，也有房仲找投資客來看房，但大家都覺得這房子太邪門，認為根本是「抓交替」。後來有位投資客表示意願想買，但到簽約當天屋主
卻反悔，表示不想讓對方成為上吊身亡的「第七個人」，所以直到現在，那間房子依然空在那裡，也沒有其他房客入住。
★ 三立新聞網提醒您：勇敢求救並非弱者，生命一定可以找到出路。透過守門123步驟-1問2應3轉介，你我都可以成為自殺防治守門人。
※ 安心專線：0800-788-995（0800-請幫幫-救救我）
※ 張老師專線：1980
※ 生命線專線：1995
※ 反霸凌專線：0800-200-885
有效內容率: 2.16%
---------------------------------------------------------------------------
```

## 深入了解

* 新聞 console 工具
* 新聞 API
* 股市交易資訊爬蟲


docs/changelog.md
## 0.3.x

0.3 的開發主軸是財經資料蒐集，針對證交所，櫃買中心，集保中心，公開資訊觀測站等地方取得有價值資料，並且匯入到 SQLite 方便使用。

### 0.3.3

* 櫃買中心三大法人，融資融券，鉅額交易 - #67
* 證交所爬蟲穩定性改善 - #69
* 集保中心爬蟲 Log Level 調整 - #73
* LZMA 壓縮快取檔 - #77 #78 #79
* 依新聞台名稱分配快取目錄 - #76
* 處理蘋果日報亂碼問題 - #80 #81
* 事務性工作 - #85 #74 #70

### 0.3.2

* 爬證交所三大法人，信用交易，借券賣出，鉅額交易，ETF折溢價率 - #56 #68
* 股權分散表爬蟲重構 - #66
* 修正新聞爬蟲對自由天下與自由時尚的相容問題 - #61
* 修正蘋果日報內文無法取得問題 - #62
* 修正東森時尚日期格式問題 - #63
* 事務性工作 - #72

### 0.3.1

* 股權分散表改用等級分表 - #57
* 股權分散表日期改用 ISO 格式 - #58
* 修正蘋果地產無法使用 https 問題 - #54
* 修正中時網頁結構改變問題 - #59
* 事務性工作 - #60

### 0.3.0

* 爬股權分散表 - #51
* 解決 // 開頭網址無法正確處理的問題 - #50 (回報者: [CpOuyang](//github.com/CpOuyang))
* 東森測試程式增加 proxy 參數解決檔日本 linode 問題 - #52
* 事務性工作 - #55

## 0.2.x

0.2 的開發主軸是新聞搜尋，針對各家新聞網站的搜尋功能，設計成行為一致的 API，簡化找新聞的需求，也解決拆新聞過程中發現的問題，列入測試樣本。

### 0.2.5

* Issue Template - #46

### 0.2.4

* 搜尋功能測試程式 - #37
* 增加 proxy 設計，讓東森新聞雲在日本 linode 也能使用 - #27
* 解決自由時報存在多種日期格式問題 - #34
* 強化記者姓名的辨識能力 - #40
* 被自動轉址時，自動切換分解新聞設定 - #43
* 修正蘋果日報搜尋的翻頁參數錯誤 - #44
* 事務性工作 - #45

### 0.2.3

* 解決各家新聞的子頻道編排相容性問題 - #31
* 新聞爬蟲重構 - #39
* 程式碼導入 pylint 檢查，遵循 PEP8 規範 - #38
* 事務性工作 - #42

### 0.2.2

* 修正 isinstance 用錯類別的問題 - #38

### 0.2.1

* 新聞標題統計功能 - #33
* 蘋果日報搜尋改用 XHR - #30
* 事務性工作 - #32

### 0.2.0

* 新聞關鍵字搜尋功能 - #9

# 0.1.x

0.1 的開發主軸是設計一個通用型的爬蟲，透過爬蟲設定值依照不同的新聞網站拆解 HTML 結構，取得標題，內文，記者，日期等資訊。

### 0.1.10

* 解決 Python 3.7 for Windows 相容性問題 - #28
* 解決 Python 3.5 3.6 for Windows 相容性問題 - #29

### 0.1.9

* 作者欄位加強測試 - #7
* 不理想 URL 測試 - #10
* 壓縮快取檔 - #15
* 壓縮範本檔 - #16
* 使用檔案儲存 log - #17
* 提供 URL 參數直接在 console 拆新聞 - #20
* 新聞爬蟲定期測試程式 - #24
* Python 3.5 相容設計 - #25
* 事務性工作 - #26

### 0.1.8

* 爬蟲支援中時電子報 - #4
* 事務性工作 - #12 #13 #14

### 0.1.7

* 清除快取前，先偵測目錄是否存在 - #2
* 變更快取目錄 - #5
* 移除 unicode_escape() - #6
* 計算有效內容率 - #8
* 變更頻果日報標題 CSS Selector - #11
* 事務性工作 - #3


twnews/soup.py
"""
新聞分解模組
"""

import io
import re
import copy
import lzma
import hashlib
import os
import os.path
from datetime import datetime

import requests
import requests.exceptions
from bs4 import BeautifulSoup

import twnews.common

def get_cache_filepath(channel, uri):
    """
    取得快取檔案路徑
    """
    cache_id = hashlib.md5(uri.encode('ascii')).hexdigest()
    path = '{}/{}.html.xz'.format(twnews.common.get_cache_dir(channel), cache_id)
    return path

def url_follow_redirection(url, proxy_first):
    """
    取得轉址後的 URL
    """
    logger = twnews.common.get_logger()
    session = twnews.common.get_session(proxy_first)
    old_url = url
    new_url = ''
    done = False

    while not done:
        try:
            resp = session.head(old_url)
            status = resp.status_code
            if status in (301, 302):
                dest = resp.headers['Location']
                if dest.startswith('//'):
                    new_url = 'https:' + dest
                elif dest.startswith('/'):
                    new_url = old_url[0:old_url.find('/', 10)] + dest
                else:
                    new_url = dest
                logger.debug('===== 轉址細節 =====')
                logger.debug('HTTP Status: %d', status)
                logger.debug('Location: %s', dest)
                logger.debug('原始 URL: %s', old_url)
                logger.debug('變更 URL: %s', new_url)
                logger.debug('====================')
                old_url = new_url
            elif status == 200:
                done = True
            else:
                logger.error('檢查轉址過程發生錯誤')
                logger.error('HTTP Status: %d，', status)
                logger.error('URL: %s，', old_url)
                done = True
        except requests.exceptions.ConnectionError as ex:
            logger.error('檢查轉址過程連線失敗: %s', ex)
            done = True

    return old_url

def url_force_https(url):
    """
    強制使用 https
    """
    logger = twnews.common.get_logger()
    if url.startswith('http://'):
        new_url = 'https://' + url[7:]
        logger.debug('原始 URL: %s', url)
        logger.debug('變更 URL: %s', new_url)
    else:
        new_url = url

    # 蘋果地產例外，要喬回來
    if new_url.startswith('https://home.appledaily.com.tw'):
        new_url = 'http://home.appledaily.com.tw' + new_url[30:]

    return new_url

def url_force_ltn_mobile(url):
    """
    強制使用自由時報行動版
    """
    logger = twnews.common.get_logger()
    new_url = url
    sub_chanels = r'^https://(3c|auto|ec|ent|food|istyle|market|playing|sports).ltn.com.tw/[^m].+'
    if url.startswith('https://news.ltn.com.tw'):
        new_url = 'https://m.ltn.com.tw' + url[len('https://news.ltn.com.tw'):]
        logger.debug('原始 URL: %s', url)
        logger.debug('變更 URL: %s', new_url)
    elif re.match(sub_chanels, url):
        uri_pos = url.find('/', 10)
        new_url = url[:uri_pos] + '/m' + url[uri_pos:]
        logger.debug('原始 URL: %s', url)
        logger.debug('變更 URL: %s', new_url)

    return new_url

def soup_from_website(url, channel, refresh, proxy_first):
    """
    網址轉換成 BeautifulSoup 4 物件
    """
    logger = twnews.common.get_logger()
    session = twnews.common.get_session(proxy_first)

    # 嘗試使用快取
    soup = None
    rawlen = 0
    uri = url[url.find('/', 10):]
    path = get_cache_filepath(channel, uri)
    if os.path.isfile(path) and not refresh:
        logger.debug('發現快取, URL: %s', url)
        logger.debug('載入快取, PATH: %s', path)
        (soup, rawlen) = soup_from_file(path)

    # 下載網頁
    if soup is None:
        logger.debug('GET URL: %s', url)
        try:
            resp = session.get(url, allow_redirects=False)
            if resp.status_code == 200:
                logger.debug('回應 200 OK')
                if resp.headers['content-type'].find('charset=') == -1:
                    # #80: 蘋果日報主頻道 Content-Type: text/html
                    #      因為沒有指定編碼造成 requests 誤判, 使用小伎倆迴避
                    content = resp.content.decode('utf-8')
                else:
                    content = resp.text
                soup = BeautifulSoup(content, 'lxml')
                rawlen = len(resp.text.encode('utf-8'))
                with lzma.open(path, 'wt') as cache_file:
                    logger.debug('寫入快取: %s', path)
                    cache_file.write(content)
            else:
                logger.warning('回應碼: %d', resp.status_code)
        except requests.exceptions.ConnectionError as ex:
            logger.error('連線失敗: %s', ex)

    return (soup, rawlen)

def soup_from_file(file_path):
    """
    本地檔案轉換成 BeautifulSoup 4 物件
    """
    html = None
    soup = None
    clen = 0

    if file_path.endswith('.xz'):
        with lzma.open(file_path, 'rt') as cache_file:
            html = cache_file.read()
    else:
        with open(file_path, 'rt') as cache_file:
            html = cache_file.read()

    if html is not None:
        soup = BeautifulSoup(html, 'lxml')
        clen = len(html.encode('utf-8'))

    return (soup, clen)

def scan_author(article):
    """
    從新聞內文找出記者姓名
    """

    patterns = [
        (r'(記者|中心)(\w{2,5})[/／╱](.+報導|特稿)', 2),
        (r'文[/／╱]記者(\w{2,5})', 1),
        (r'[\(（](\w{2,5})[/／╱].+報導[\)）]', 1),
        (r'記者(\w{2,3}).{2}[縣市]?\d{1,2}日電', 1),
        (r'(記者|遊戲角落 )(\w{2,5})$', 2),
        (r'\s(\w{2,5})[/／╱].+報導$', 1),
        (r'（譯者：(\w{2,5})/.+）', 1),
        (r'【(\w{2,5})╱.+報導】', 1)
    ]

    exclude_list = [
        '國際中心',
        '地方中心',
        '社會中心',
        '攝影'
    ]

    for (patt, gidx) in patterns:
        pobj = re.compile(patt)
        match = pobj.search(article)
        if match is not None:
            if match.group(1) not in exclude_list:
                return match.group(gidx)

    return None

class NewsSoup:
    """
    新聞分解器
    """

    # pylint: disable=too-many-instance-attributes

    def __init__(self, path, refresh=False, proxy_first=False):
        """
        建立新聞分解器
        """

        self.path = path
        self.refresh = refresh
        self.proxy_first = proxy_first
        self.loaded = False
        self.soup = None
        self.rawlen = 0
        self.logger = twnews.common.get_logger()
        self.channel = twnews.common.detect_channel(path)
        self.cache = {
            'title': None,
            'date_raw': None,
            'date': None,
            'author': None,
            'contents': None,
            'tags': None
        }

        if self.channel == '':
            self.logger.error('不支援的新聞台，請檢查設定檔')
            return

        # URL 正規化
        if self.path.startswith('http'):
            self.path = url_follow_redirection(self.path, self.proxy_first)
            self.path = url_force_https(self.path)
            if self.channel == 'ltn':
                self.path = url_force_ltn_mobile(self.path)

        # Layout 偵測
        layout = 'mobile'
        layout_list = twnews.common.get_channel_conf(self.channel, 'layout_list')
        for item in layout_list:
            if self.path.startswith(item['prefix']):
                layout = item['layout']

        self.conf = twnews.common.get_channel_conf(self.channel, layout)

    def __get_soup(self):
        if not self.loaded:
            self.loaded = True
            try:
                if self.path.startswith('http'):
                    self.logger.debug('從網路載入新聞')
                    (self.soup, self.rawlen) = soup_from_website(
                        self.path,
                        self.channel,
                        self.refresh,
                        self.proxy_first
                    )
                else:
                    self.logger.debug('從檔案載入新聞')
                    (self.soup, self.rawlen) = soup_from_file(self.path)
            except requests.ConnectionError as ex:
                self.logger.error('因連線問題，無法載入新聞: %s', ex)
                self.logger.error(self.path)
            except FileNotFoundError as ex:
                self.logger.error('檔案不存在，無法載入新聞: %s', ex)
                self.logger.error(self.path)
            except TypeError as ex:
                self.logger.error('頻道不存在，無法載入新聞: %s', ex)
                self.logger.error(self.path)

            if self.soup is None:
                self.logger.error('無法轉換 BeautifulSoup，可能是網址或檔案路徑錯誤')

        return self.soup

    def title(self):
        """
        取得新聞標題
        """

        soup = self.__get_soup()
        if soup is None:
            return None

        if self.cache['title'] is None:
            nsel = self.conf['title_node']
            found = soup.select(nsel)
            if found:
                node = copy.copy(found[0])
                # 避免子元件干擾日期格式
                for child_node in node.select('*'):
                    child_node.extract()
                self.cache['title'] = node.text.strip()
                if len(found) > 1:
                    self.logger.warning('找到多組標題節點 (新聞台: %s)', self.channel)
            else:
                self.logger.error('找不到標題節點 (新聞台: %s)', self.channel)

        return self.cache['title']

    def date_raw(self):
        """
        取得原始時間字串
        """

        soup = self.__get_soup()
        if soup is None:
            return None

        if self.cache['date_raw'] is None:
            nsel = self.conf['date_node']
            found = soup.select(nsel)
            if found:
                node = copy.copy(found[0])
                # 中時: 日期散落在兩個子節點，不可丟棄子節點
                # 聯合: 日期在這個節點，子節點有其他文字，必須丟棄子節點
                if 'date_with_children' not in self.conf or not self.conf['date_with_children']:
                    # 丟棄子節點
                    for child_node in node.select('*'):
                        child_node.extract()
                self.cache['date_raw'] = node.text.strip()
                if len(found) > 1:
                    self.logger.warning('發現多組日期節點 (新聞台: %s)', self.channel)
            else:
                self.logger.error('找不到日期時間節點 (新聞台: %s)', self.channel)

        return self.cache['date_raw']

    def date(self):
        """
        取得 datetime.datetime 格式的時間
        """

        soup = self.__get_soup()
        if soup is None:
            return None

        if self.cache['date'] is None:
            formats = self.conf['date_format']
            if isinstance(formats, str):
                formats = [formats]

            for dfmt in formats:
                try:
                    self.cache['date'] = datetime.strptime(self.date_raw(), dfmt)
                except TypeError as ex:
                    errmsg = '日期格式分析失敗 {} (新聞台: {})'.format(ex, self.channel)
                except ValueError as ex:
                    errmsg = '日期格式分析失敗 {} (新聞台: {})'.format(ex, self.channel)

            if self.cache['date'] is None:
                self.logger.error(errmsg)

        return self.cache['date']

    def author(self):
        """
        取得新聞記者/社論作者
        """

        # pylint: disable=too-many-branches

        soup = self.__get_soup()
        if soup is None:
            return None

        if self.cache['author'] is None:
            nsel = self.conf['author_node']
            if nsel != '':
                if isinstance(nsel, str):
                    selectors = [nsel]
                else:
                    selectors = nsel
                for nsel in selectors:
                    found = soup.select(nsel)
                    if found:
                        node = copy.copy(found[0])
                        for child_node in node.select('*'):
                            child_node.extract()
                        author_raw = node.text.strip()
                        if author_raw[0] != '記' and len(author_raw) <= 5:
                            self.cache['author'] = author_raw
                        else:
                            self.cache['author'] = scan_author(author_raw)
                        if len(found) > 1:
                            self.logger.warning('找到多組記者姓名 (新聞台: %s)', self.channel)
                        break
                    else:
                        self.logger.warning('找不到記者節點 (新聞台: %s)', self.channel)
            else:
                contents = self.contents()
                if contents is not None:
                    self.cache['author'] = scan_author(contents)
                    if self.cache['author'] is None:
                        self.logger.warning('內文中找不到記者姓名 (新聞台: %s)', self.channel)
                else:
                    self.logger.error('因為沒有內文所以無法比對記者姓名 (新聞台: %s)', self.channel)

        return self.cache['author']

    def contents(self, limit=0):
        """
        取得新聞內文
        """

        soup = self.__get_soup()
        if soup is None:
            return None

        if self.cache['contents'] is None:
            nsel = self.conf['article_node']
            found = soup.select(nsel)
            if found:
                contents = io.StringIO()
                for node in found:
                    contents.write(node.text.strip())
                self.cache['contents'] = contents.getvalue()
                contents.close()
            else:
                self.logger.error('找不到內文節點 (新聞台: %s)', self.channel)

        if isinstance(self.cache['contents'], str) and limit > 0:
            # https://github.com/PyCQA/pylint/issues/1498
            # pylint: disable=unsubscriptable-object
            return self.cache['contents'][0:limit]

        return self.cache['contents']

    def effective_text_rate(self):
        """
        計算有效內容率 (有效內容位元組數/全部位元組數)
        """

        soup = self.__get_soup()
        if soup is None or self.rawlen == 0:
            return 0

        data = [
            self.title(),
            self.author(),
            self.date_raw(),
            self.contents()
        ]

        useful_len = 0
        for datum in data:
            if datum is not None:
                useful_len += len(datum.encode('utf-8'))

        return useful_len / self.rawlen
